### PR TITLE
SRCH-3276 added uswds banner and footer to login and signup pages

### PR DESCRIPTION
**Pull request by yogesh27**
## Summary
- Added the css and js for USWDS banner and footer components
- Added the banner and footer components
- Added the Cucumber tests
- Not using the vendor's uswds js, wrote the custom code in usdws.js to handle the banner toggle functionality
 
### Checklist
Please ensure you have addressed all concerns below before marking a PR "ready for review" or before requesting a re-review. If you cannot complete an item below, replace the checkbox with the ⚠️ `:warning:` emoji and explain why the step was not completed.
 
#### Functionality Checks

- [x] You have run `bundle update` and committed your changes to Gemfile.lock.
 
- [x] You have merged the latest changes from the target branch (usually `main`) into your branch.
 
- [x] You have squashed your commits into a single commit (exceptions: your PR includes commits with formatting-only changes, such as required by Rubocop or Cookstyle, or if this is a feature branch that includes multiple commits).
 
- [x] Your primary commit message is of the format **SRCH-#### \<description\>** matching the associated Jira ticket.

- [x] PR title is either of the format **SRCH-#### \<description\>** matching the associated Jira ticket (i.e. "SRCH-123 implement feature X"), or **Release - SRCH-####, SRCH-####, SRCH-####** matching the Jira ticket numbers in the release.
 
- [ ] Automated checks pass. If Code Climate checks do not pass, explain reason for failures:
 
#### Process Checks

- [x] You have specified at least one "Reviewer".

### CODE REVIEW

**Comment by yogesh27** [comment]
Code Climate failing due to CSS linting. All the CSS is from the USWDS vendor, so we should be good.

**Comment by MothOnMars** [comment]
@yogesh27 , you can tell CodeClimate to ignore the imported CSS files:
```yaml
#.codeclimate.yml

exclude_patterns:
...
# These files were downloaded from https://designsystem.digital.gov/
- app/assets/stylesheets/uswds.css
- app/assets/stylesheets/uswds_override.css

**Comment by yogesh27** [comment]
> @yogesh27 , you can tell CodeClimate to ignore the imported CSS files:
> 
> ```yaml
> #.codeclimate.yml
> 
> exclude_patterns:
> ...
> # These files were downloaded from https://designsystem.digital.gov/
> - app/assets/stylesheets/uswds.css
> - app/assets/stylesheets/uswds_override.css
> ```

Thank you. I have updated the codeclimate.yml

**Review by MothOnMars — CHANGES_REQUESTED** [review]
@yogesh27 , I took a quick look at this PR (leaving the main review to @danswick ). A couple of comments:
- be sure to keep the PR in "draft" mode until all automated checks are passing (with the exception of Code Climate)
- please add tests for this, especially for the custom JS. We don't have a JS testing framework, but you should be able to test with Cucumber (I'm happy to pair if you haven't worked with Cucumber in the past).

**Comment by yogesh27** [comment]
> 

Sure, Sounds good.

**Comment by yogesh27** [comment]
All the checks are now passing.

**Comment by yogesh27** [comment]
@MothOnMars I have added the cucumber tests for the header banner and footer section for login and signup pages.

File: app/assets/javascripts/uswds.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+var button = document.getElementsByClassName("usa-banner__button")[0];
+var banner_content = document.getElementsByClassName("usa-banner__content")[0];
+
+button.addEventListener("click", function(){
+    var is_hidden = banner_content.getAttribute("hidden");
```
Comment (MothOnMars) [inline app/assets/javascripts/uswds.js:5]: @yogesh27 , our existing JS files use two spaces for indentation in JS files. Unless you feel strongly that we should use 4, I think we should make 2 the default (let's discuss general JS linting w/ the team at scrum). Our Ruby files all use 2 spaces, which seems to be turning into the standard for JS, too: https://www.w3schools.com/js/js_conventions.asp

File: features/user_sessions.feature [hunk 2]
```diff
@@ -46,3 +46,66 @@ Feature: User sessions
     When I sign out
     And I go to the usagov's Dashboard page
     Then I should see "Security Notification"
+  
+  @javascript
+  Scenario: User sees the header with USWDS banner
+    When I go to the login page
+    Then I should see "An official website of the United States government"
```
Comment (MothOnMars) [inline features/user_sessions.feature:53]: Please add `And I should not see "Official websites use .gov"` before the button press; this confirms that the partial is hidden until the user clicks the button.

File: features/user_sessions.feature [hunk 3]
```diff
@@ -46,3 +46,66 @@ Feature: User sessions
...
+    When I press "Here’s how you know"
+    Then I should see "Official websites use .gov"
+    And I should see "A .gov website belongs to an official government organization in the United States."
+    And I should see "Secure .gov websites use HTTPS"
+    And I should see "A lock"
+    And I should see "https:// means you’ve safely connected to the .gov website. Share sensitive information only on official, secure websites."
```
Comment (MothOnMars) [inline features/user_sessions.feature:59]: I would simplify this and the other scenarios by checking just a single string in the relevant partial. That keeps the features from being bloated and becoming unreadable. No need to check multiple elements in each partial, unless there's a specific need.

File: app/assets/javascripts/uswds.js (hunk 1 above)
Comment (yogesh27) [reply app/assets/javascripts/uswds.js:5]: We can use 2 spaces. I will update the editor settings.

File: features/user_sessions.feature (hunk 2 above)
Comment (yogesh27) [reply features/user_sessions.feature:53]: Sure, Sounds good.

File: features/user_sessions.feature (hunk 3 above)
Comment (yogesh27) [reply features/user_sessions.feature:59]: Sure. In the footer, I am checking for the all links and their text as well. Do we need that? or I can just check couple of links?

File: features/user_sessions.feature [hunk 4]
```diff
@@ -46,3 +46,66 @@ Feature: User sessions
     When I sign out
     And I go to the usagov's Dashboard page
     Then I should see "Security Notification"
+  
+  @javascript
```
Comment (MothOnMars) [inline features/user_sessions.feature:50]: Use the `@javascript` tag sparingly, as it uses a slower driver. The only scenario that should need it is the one testing the dropdown functionality in the header.

File: features/user_sessions.feature [hunk 5]
```diff
@@ -46,3 +46,66 @@ Feature: User sessions
...
+    And I should see a link to "https://gsa.gov/tts" with text "Technology Transformation Services"
+    And I should see a link to "https://www.gsa.gov/about-us" with text "About GSA"
+    And I should see a link to "https://www.gsa.gov/website-information/accessibility-aids" with text "Accessibility support"
+    And I should see a link to "https://www.gsa.gov/reference/freedom-of-information-act-foia" with text "FOIA requests"
+    And I should see a link to "https://www.gsa.gov/reference/civil-rights-programs/notification-and-federal-employee-antidiscrimination-and-retaliation-act-of-2002" with text "No FEAR Act data"
+    And I should see a link to "https://www.gsaig.gov/" with text "Office of the Inspector General"
+    And I should see a link to "https://www.gsa.gov/reference/reports/budget-performance" with text "Performance reports"
+    And I should see a link to "https://www.gsa.gov/website-information/website-policies" with text "Website policies"
+    And I should see "Looking for U.S. government information and services?"
+    And I should see a link to "https://www.usa.gov/" with text "Visit USA.gov"
```
Comment (MothOnMars) [inline features/user_sessions.feature:111]: I recommend consolidating these into just a couple of scenarios:
```cucumber
Scenario: User sees the Search.gov footer site links
  ...
  
Scenario: User sees the GSA footer links
  ...
```

Comment (yogesh27) [reply features/user_sessions.feature:111]: Sounds good, thanks.

File: features/user_sessions.feature (hunk 3 above)
Comment (MothOnMars) [reply features/user_sessions.feature:59]: Yeah, I would check a minimal amount of content just to ensure that each partial appears. See comment below.

**Comment by yogesh27** [comment]
> 

Yes, it started to fail soon after I did the bundle upgrade. Let me downgrade it and check. Thanks

**Comment by MothOnMars** [comment]
> > 
> 
> Yes, it started to fail soon after I did the bundle upgrade. Let me downgrade it and check. Thanks

My branch using webmock 3.18.1 passed, so I'm not sure what the issue is. You could also try backing out all of the gem upgrades. I'll see if a clean branch with an updated gemset passes.

**Comment by MothOnMars** [comment]
@yogesh27 , actually, run `bundle update` again to bump Webmock to 3.18.1. They just fixed the issue we've been seeing:
https://github.com/bblimke/webmock/issues/999

**Comment by yogesh27** [comment]
All the checks are now passing. Using webmock 3.18.1 and ruby as platform in Gemfile.lock

**Comment by MothOnMars** [comment]
@yogesh27 , the new cucumber tests look great. I'll leave the full review to @danswick , but so far, LGTM!

File: app/assets/javascripts/uswds.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+  var is_hidden = banner_content.getAttribute("hidden");
+  if(is_hidden){
+    banner_content.removeAttribute("hidden");
+    this.setAttribute("aria-expanded", "true");
+  }
+  else{
+    banner_content.setAttribute("hidden", "true");
+    this.setAttribute("aria-expanded", "false");
+  }
+});
```
Comment (danswick) [inline app/assets/javascripts/uswds.js:14]: Totally up to your preferences, but this made me think of `toggleAttribute`: 

```suggestion
button.addEventListener("click", function(){
  banner_content.toggleAttribute("hidden");
  this.setAttribute("aria-expanded",
    this.getAttribute("aria-expanded") === "true"
      ? "false"
      : "true"
  );
});
```

**Review by danswick — APPROVED** [review]
This looks good to me, @yogesh27. I left one comment for you to consider (totally up to you if you want to take it). Also wanted to note that I did not build and test locally, but am happy to do so if it would be helpful!

File: app/assets/javascripts/uswds.js (hunk 6 above)
Comment (yogesh27) [reply app/assets/javascripts/uswds.js:14]: Thanks @danswick - this is a good suggestion. I will update my code accordingly.

Comment (yogesh27) [reply app/assets/javascripts/uswds.js]: Hi @danswick I have taken care of your suggestions.

**Review by danswick — APPROVED** [review]
Your updates look good to me, @yogesh27 👍

**Comment by yogesh27** [comment]
Hi @danswick While testing the code on staging, some css for fav icon for arrow up/down next to "Here’s how you know" was not working, so I have made some css fixes for it.

File: app/assets/stylesheets/uswds.css [hunk 7]
```diff
@@ -0,0 +1,4048 @@
...
+  .usa-banner__button::after{
+    background:url("uswds/usa-icons/expand_more.svg") no-repeat center/contain;
+    display:inline-block;
+    height:1rem;
+    width:1rem;
+    content:"";
+    vertical-align:middle;
+    margin-left:2px;
+  }
+  /* @supports ((-webkit-mask: url("")) or (mask: url(""))){ */
```
Comment (danswick) [inline app/assets/stylesheets/uswds.css]:
```suggestion
  /* at-rule removed to fix a display issue on the search site. Commented lines retained for future reference if this file needs to be updated from USWDS
  /* @supports ((-webkit-mask: url("")) or (mask: url(""))){ */
```

**Review by danswick — APPROVED** [review]
Your CSS updates look good to me, @yogesh27 👍

File: app/assets/stylesheets/uswds.css (hunk 7 above)
Comment (yogesh27) [reply app/assets/stylesheets/uswds.css]: Comments have been removed.

**Comment by yogesh27** [comment]
Hi @danswick I made some updates based on the @amanifarooque 's feedback. Kindly review.

**Comment by yogesh27** [comment]
Hi @danswick Kindly review the minor changes I did to incorporate updated footer links

**Review by danswick — APPROVED** [review]
Updated footer items look good to me, @yogesh27 👍

**Pull request merged**